include/openrave/posturedescriber.h

```diff
@@ -0,0 +1,115 @@
+    /// \return true if this describer class can support the posture description AND the initialization is successful.
+    virtual bool Init(const RobotBase::ManipulatorPtr& pmanip);
+
+    /// \brief Computes posture state integers for a kinematics chain at either the current of specified dof values.
+    /// \param [in]  dofvalues       if empty, then use the current dof values; otherwise these specified dof values must have the same size as the number of dofs in the kinematics chain.
+    /// \param [out] posturestates   posture states, whose size is a power of 2. Always non-empty if (1) this class is properly initialized AND (2) dofvalues is either empty or has the correct size.
+    /// \return true if (1) this describer class is properly initialized AND (2) dofvalues is either empty or has the correct size.
+    virtual bool ComputePostureStates(std::vector<PostureStateInt>& posturestates,
+                                     const std::vector<double>& dofvalues = {},
+                                     const std::vector<int>& dofindices = {},
+                                     const KinBody::CheckLimitsAction claoption = KinBody::CheckLimitsAction::CLA_Nothing
+                                     ) = 0;
+
+    /// \brief Gets the key used in map data (of type CustomData) in IkReturn
+    virtual std::string GetMapDataKey() const = 0;
+
+    /// \brief Checks if the posture describer has been initialzied
+    virtual bool IsInitialized() const = 0;
+
+    /// \brief Cleans internal setup in the early stage of doing initializaion by Init()
+    virtual void Destroy() = 0;
+
+    /// \brief Gets essential kinematics chain associated with this describer
+    virtual const LinkPair& GetEssentialKinematicsChain() const = 0;
+
+    /// \return the static interface type this class points to (used for safe casting)
+    static InterfaceType GetInterfaceTypeStatic() {
+        return PT_PostureDescriber;
+    }
+
+private:
+    virtual const char* GetHash() const final;
+};
+
+using PostureDescriberBasePtr = boost::shared_ptr<PostureDescriberBase>;
+
+///< \brief Acquires from the manipulator the essential kinematics chain in the form of a link pair.
+/// An essential kinematics chain, if not empty, is always such that both the first joint and the last joint are active and revolute.
+OPENRAVE_API LinkPair ExtractEssentialKinematicsChain(const LinkPair& kinematicsChain);
+OPENRAVE_API LinkPair ExtractEssentialKinematicsChain(const RobotBase::ManipulatorPtr& pmanip);
+OPENRAVE_API LinkPair ExtractEssentialKinematicsChain(const RobotBase::ManipulatorConstPtr& pmanip);
+
+///< \brief Acquires from the manipulator the baselink-eelink pair.
+OPENRAVE_API LinkPair GetKinematicsChain(const RobotBase::ManipulatorConstPtr& pmanip);
+
+///< \brief Computes a kinematics hash from the baselink to eelink.
+OPENRAVE_API std::string ComputeKinematicsChainHash(const LinkPair& kinematicsChain, std::vector<int>& armindices);
```

Comment:
let's remove ``armindices`` from this function's parameter

Reply:
done